tests/sources/test_redis.py

```diff
@@ -0,0 +1,370 @@
+            ),
+        ),
+        (
+            # valid: one custom pattern
+            [{"database": 0, "key_pattern": "*"}],
+            SyncRuleValidationResult.valid_result(
+                SyncRuleValidationResult.ADVANCED_RULES
+            ),
+        ),
+        (
+            # valid: two custom patterns
+            [
+                {"database": 0, "key_pattern": "test*"},
+                {"database": 1, "type": "string"},
+            ],
+            SyncRuleValidationResult.valid_result(
+                SyncRuleValidationResult.ADVANCED_RULES
+            ),
+        ),
+        (
+            # invalid: database number
+            [{"database": -1}],
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+        (
+            # invalid: array of arrays -> wrong type
+            {"database": ["a/b/c", ""]},
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+        (
+            # invalid database name
+            {"database": 0, "key_pattern": "abc*"},
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+        (
+            # invalid: key_pattern or type is missing
+            {"database": 0},
+            SyncRuleValidationResult(
+                SyncRuleValidationResult.ADVANCED_RULES,
+                is_valid=False,
+                validation_message=ANY,
+            ),
+        ),
+    ],
+)
+@pytest.mark.asyncio
+async def test_advanced_rules_validation(advanced_rules, expected_validation_result):
+    async with create_redis_source() as source:
+        source.client._client = RedisClientMock()
+        validation_result = await RedisAdvancedRulesValidator(source).validate(
+            advanced_rules
+        )
+        assert validation_result == expected_validation_result
+
+
+@pytest.mark.asyncio
+async def test_client_when_mutual_ssl_enabled():
+    async with create_redis_source() as source:
+        source.client.database = ["*"]
+        source.client.ssl_enabled = True
+        source.client.mutual_tls_enabled = True
+        source.client.tls_certfile = (
+            "-----BEGIN CERTIFICATE----- Invalid-Certificate -----END CERTIFICATE-----"
+        )
+        source.client.tls_keyfile = "-----BEGIN RSA PRIVATE KEY----- Invalid-Certificate -----END RSA PRIVATE KEY-----"
+        mocked_client = AsyncMock()
+        with mock.patch(
+            "redis.from_url",
+            return_value=mocked_client,
+        ):
+            redis_client = source.client._client
+            assert hasattr(redis_client, "ping")
```

Comment:
Will this pass? `source.client._client` (i.e. an `AsyncMock` instance) was assigned to `redis_client`, and I don't think it has the attribute `ping`.

Reply:
The expectation of this test case (when mutual ssl is enabled) is that:
1. You mock `redis.from_url`
2. You have the source configured with `mutual_tls_enabled = Ture`
3. The mocked `reids.from_url` should be called with the expected url, i.e. rediss + query

Same for the test case when ssl is enabled, it should expect a call with rediss